### PR TITLE
59 virtual server selection target resource location

### LINKED ISSUE #59: Virtual server selection / target resource location
### Virtual Server selection

Add functionality to select the correct virtual server, based on host/port and Host request header.

Server can be set in different stages:
- Idle: right after connection pick first fitting Host/port combination based on serverSock of connection
- ReceiveHeader: after request header has been parsed check if Host field gives new information

Will add a pointer to a ConfigServer in HTTPRequest.

### Target resource location

Right now the location of the target resource is found in the ResponseBuilder. This is too late since config values of the location are needed earlier (ReceiveBody). 
Should be done after parsing request headers

Will add a pointer to a Location in HTTPRequest.

### References
- [How nginx processes a request](https://nginx.org/en/docs/http/request_processing.html)
- [Virtual server selection](https://nginx.org/en/docs/http/server_names.html#virtual_server_selection)
- [Understanding Nginx Server and Location Block Selection Algorithms](https://www.digitalocean.com/community/tutorials/understanding-nginx-server-and-location-block-selection-algorithms)

### PR DESCRIPTION
**Pull request by gwolf-011235**
### Summary
1. Select server configuration for connection
2. Write own implementation of getnameinfo()
3. Use gethostname()
4. Refactor: find TargetResource in connectionReceiveHeader
5. Refactor: delete HTTPResponse, use HTTPRequest instead
6. Misc

### 1. Select server configuration for connection
- new function hasValidServerConfig() : compares serverSocket of connection with configs in ConfigFile and finds a match. Match is if host and port are equal. If more than one matches are found, selects first match.
- If no matches found, tries to match with a wildcard server (right now only IPv4, i.e. 0.0.0.0). If one is present, it only needs to match port.
- Overload: if additionally a host string is passed, also checks server_name of found matches
- Connection::Connection: when constructing a new Connection, vector of servers need to be passed. Ctor of connection does this check (aka selects a default server)
- connectionReceiveHeader: after receiving a complete request header, checks if Host header field is present. If it is, checks server config again
- imitates nginx, see also resources in #59

### 2. Write own implementation of getnameinfo()
- formerly used getnameinfo() to parse struct addr* > getting host and port as string. Turns out this function is not allowed (I think I looked at older version of subject)
- Implemented own version, but only for IPv4.
- If we want to support IPv6, we need to implement additional algorithm 

### 3. Use gethostname()
- when a new Connection is constructed, it gets clientSock (representing client host:port) and serverSock (representing server host:port)
- if a wildcard server (0.0.0.0) would accept() the connection, the wildcard host would be saved, not the actual host.
- to find out actual host (e.g. 127.0.0.1) gethostname() is used
- double checked if function is allowed in subject.
- is needed to select correct server config

### 4. Refactor: find TargetResource in connectionReceiveHeader
- we need info about target ressource (or more precise: about target location) earlier than in ResponseBuilder: e.g. maxbodysize can be set in location
- moved TargetResourceHandler into Server. It is called in connectionReceiveHeader, after complete request header was received.
- needed to refactor to adapt to new place in code
- also refactored connectionReceiveHeader() by splitting of handleCompleteRequestHeader()
- adapted tests

### 5. Refactor: delete HTTPResponse, use HTTPRequest instead
- HTTPResponse was only used in ResponseBuilder and friends.
- Much of it was just copied over from HTTPRequest.
- Supplemented stuff to HTTPRequest and deleted HTTPResponse entirely
- Refactored ResponseBuilder and friend to directly use HTTPRequest
- adapted tests

### 6. Misc
- adapted error msg in FileSystemPolicy; AutoIndexHandler; TargetResourceHandler to use logger
- moved timestamp generation in connectionReceiveHeader and connectionReceiveBody higher up, adapted tests
- clear() connection buffer in connectionBuildResponse
- add constructor for HTTPRequest to set default values

closes #59

### CODE REVIEW

**Review by u413-284-si — COMMENTED** [review]
Only a couple of questions, nothing major popped up. 
Incorporated changes into CGIHandling, so that both branches are aligned for later merge.

File: src/Server.cpp [hunk 1]
```diff
@@ -151,12 +154,14 @@ bool Server::registerVirtualServer(int serverFd, const Socket& serverSock)
  *
  * Registers a client fd with addEvent().
  * If it fails to add the event, it logs an error and closes the client socket.
- * If the connection is successfully registered, adds the connection to map m_connections.
- * The []-operator creates a new entry in the map if the key does not exist or overwrites an existing one.
+ * If the connection is successfully registered, create a new struct Connection with serverSock and clientSock.
+ * The active server of the connectionis selected with selectServerConfig().
```
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/Server.cpp]: Yes, renamed after refactor and didn't have it in the ctor before.

- Deleted the sentence "The active server of the connectionis selected with selectServerConfig()."
- Connection ctor has "Tries to find a valid server configuration for the connection with hasValidServerConfig(). [...]"

File: src/Server.cpp [hunk 2]
```diff
@@ -704,6 +752,57 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
...
+	}
+
+	server.findTargetResource(connection, connection.m_request);
+
+	if (connection.m_request.httpStatus == StatusOK && connection.m_request.hasBody) {
+		connection.m_status = Connection::ReceiveBody;
+		connection.m_buffer.erase(0, connection.m_buffer.find("\r\n\r\n") + 4);
+	} else {
+		connection.m_status = Connection::BuildResponse;
+		server.modifyEvent(clientFd, EPOLLOUT);
```
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/Server.cpp:801]: In connectionBuildResponse the connection buffer gets cleared anyways.

So we don't need to worry about cleaning it beforehand

File: src/Server.cpp [hunk 3]
```diff
@@ -752,7 +851,6 @@ void connectionReceiveBody(Server& server, int clientFd, Connection& connection)
 			}
 			connection.m_status = Connection::BuildResponse;
 			server.modifyEvent(clientFd, EPOLLOUT);
-			connection.m_buffer.erase(0);
```
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/Server.cpp:755]: See above: connectionBuildResponse always clears

**Review by QCHR1581 — APPROVED** [review]
Nothing additional from my side. 

Tests showcase that new features work as intended and the refactors make sense.

**Pull request merged**